Open paper details card in project papers view

Merseleo opened the pull request:
Closes #219

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a new button, which does x -->

- I completed the card containing the paper details in the project papers view
  - to reuse the `PaperDetailsCard`, i extracted its content to a separate svelte component that can be used for this card, as the entire `PaperDetailsCard` could not be used in the project papers view because of size and styling differences
  - I extended this card by adding two icon buttons for opening the paper or add / remove the paper to / from the reading list
  - I added an animation, so the card for the paper details slides / flies in and out
- I added an action that dispatches an event, when a user clicks outside the referring component or presses "Escape" and integrated this into the project papers view, so the "`PaperDetailsCard`" is closed when the user clicks outside this card or any other `PaperListEntry` component (when the user clicks on another `PaperListEntry` component we get a closing and reopening animation, which somehow looks weird, so I am open for suggestions)

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~~[ ] unit tests~~ (not necessary, as action was tested by integration test)
  - [x] integration tests
  - [x] end-to-end tests
- [ ] (for reviewer) I have checked the implementation against the requirements

Commit:
feat: show adapted paper details card, when clicking a paper in the 'Papers' list (= stage entry)

Commit:
feat: add icon to paper details card on project papers view that allows opening the project paper

Commit:
feat: add 'PaperBookmarkButton' to the paper details card in the project papers view (without implementation)

Commit:
feat: animates the opening / closing of the 'PaperDetailsCardContent' on the project paper view

Commit:
test: fix 'PaperBookmarkButton' tests according to updated implementation

Commit:
feat: add option to close the 'PaperDetailsCardContent' on project papers view (by pressing escape or clicking outside)

Commit:
test: add test for 'clickOutsideOrEscape' action

To test this action a dummy component using this action is created and the user interactions with this componet are simulated and checked, whether the corresponding events were fired or not

Commit:
docs: add missing words in comments

Commit:
test: add fixture for directly providing the page object models in the project creation test

Introducing a fixture for all poms related to the homepage has the advantage that whenever these poms are required or useful, you do not need to create them, but access the poms via the parameter list.

Commit:
test: add E2E test for opening and closing the paper details card on papers overview

Commit:
test: exchange css locators by getByRole locator where possible

Slartibartfass2 committed:
fix: let playwright fail when subprocess fails

mowi12 commented on src/routes/project/[projectId]/papers/+page.svelte:
```diff
@@ -143,13 +151,37 @@
+                selectedPaper = undefined;
+            }}
+            use:clickOutsideOrEscape
+            transition:fly={{ duration: 100, x: 100, opacity: 0 }}
```
```suggestion
            transition:fly={{ duration: 250, x: 150, opacity: 0 }}
```

I tried it with these values. Imo it looked better but felt a bit slow. Maybe something in between?
All the other animation types don't look as good as `fly`.

Merseleo replied:
I changed it so a duration of 200ms, what's your opinion on this change?

mowi12 replied:
Yes, looks good to me.

mowi12 commented on src/lib/components/composites/PaperBookmarkButton.svelte:
```diff
@@ -4,26 +4,32 @@
     interface Props {
         loadingPaperId: Promise<string>;
-        isBookmarkedDefault: boolean;
+        isBookmarkedDefault?: boolean;
```
Why is the modification of `isBookmarkedDefault` to `isBookmarkedDefault?` needed?

Merseleo replied:
It is not needed, but imo it removes unnecessary code. I prefer to use optional parameters when you often set a prop to the same value (as here) and only want to change it occasionally. Using optional parameters then makes the code more readable and less cluttered as you only set the values you explicitly want to set.

mowi12 replied:
You're right.

mowi12 commented on src/lib/components/composites/PaperBookmarkButton.svelte:
```diff
@@ -4,26 +4,32 @@
-        initialValue: undefined,
-        resourceName: "paper ID",
-    });
+    const tooltipText = $derived(isBookmarked ? "Remove from reading list" : "Add to reading list");
```
Why did you change the case of `Reading List`. Since it's not even a complete sentence, why not using header case or what it's called? Do we handle this similarly in other cases?

Merseleo replied:
I think we agreed to generally not use this "header style" but only capitalize the first letter.

mowi12 replied:
Okay.

mowi12 commented on tests/integration/actions/actions.test.ts:
```diff
@@ -0,0 +1,37 @@
+import ClickOutsideOrEscapeActionComponent from "./ClickOutsideOrEscapeActionComponent.svelte";
+
+describe("Click outside or press escape action", () => {
+    it("When a component is created with the 'clickOutsideOrEscape' action and the user clicks outside, then the 'ClickedOutsideOrEscape' event is fired.", async () => {
```
```suggestion
    test("When a component is created with the 'clickOutsideOrEscape' action and the user clicks outside, then the 'ClickedOutsideOrEscape' event is fired.", async () => {
```

As I understand it, there are two keywords for tests `it` and `test`. The reasons is to make a sentence with the keyword and the following sentence. So maybe use `test` instead of `it`?

This, of course, applies to all vitest tests.

Merseleo replied:
I adapted it.

mowi12 commented on tests/e2e/fixtures/general-fixture.ts:
```diff
@@ -33,7 +33,7 @@ export const defaultTest = base.extend<{ forEachTest: void } & TestOptions>({
             await use();
         },
-        { auto: true },
+        { auto: true, box: true },
```
What exactly is `box: true` needed for?

Merseleo replied:
This is not needed, but kind of useful in my opinion. Using `box: true` causes this fixture step to disappear from the UI / log (see [here](https://playwright.dev/docs/test-fixtures#box-fixtures) for more information) and as this step is always applied I think we could remove it from the log.

Merseleo replied:
But if you think we should keep this in the log for completeness I will remove it.

mowi12 replied:
No, we can definitely keep it. Makes sense now!

mowi12 commented on tests/e2e/pom/create-project-dialog-model.ts:
```diff
@@ -40,6 +40,9 @@ export class DevCreateProjectDialog {
         await this.projectMemberInput.press("Tab");
+        await this.projectMemberInput.fill("Alice");
+        await this.projectMemberInput.press("ArrowDown");
+        await this.projectMemberInput.press("Enter");
```
This is not important, but is it wanted that the user Alice has the email address of John Doe?

Merseleo replied:
It shouldn't have this email address. Actually we add two members, `john@doe.com` by email address and `Alice Smith` by name? Or did you find a but with this test?

mowi12 replied:
I had the wrong component in mind. You're right!

mowi12 reviewed (changes requested):
All in all, your code is very clean, and I couldn't find much to criticise.

I am not quite sure, but I found 2 other issue regarding filtering and searching the pages in the different stages. Currently, this functionality does not work, but in my understanding this is part of other issues (as well as adding a new paper). If this is the case, this comment can be ignored.

After fiddling around for a long time, I couldn't find a fix for the next problem. Please see to it, if you can find a way around.
As soon as you click on a paper, and the `PaperDetailsCard` appears, a horizontal scrollbar appears as well, but it's not necessary since all the content is still visible.

---

I think the following problem can be handled in this issue:
Because the value of a `PaperDetail` is displayed in a `textarea`, long titles (and maybe also other fields) can reach a length where a scrollbar appears. This looks very weird, and maybe just a line break should appear. I have no further suggestions since this can also become a separate issue since it is also effecting the paper details view of each separate paper.

<img width="441" alt="Screenshot 2025-04-20 at 08 54 47" src="https://github.com/user-attachments/assets/9d6a7220-cd3d-4410-b96b-569b1d230caf" />

Commit:
refactor: change the animation during for the paper details card from 100ms to 200ms

Commit:
test: change 'it' to 'test' keyword in all vitest tests

Merseleo reviewed (commented):
> All in all, your code is very clean, and I couldn't find much to criticise.
> 
> I am not quite sure, but I found 2 other issue regarding filtering and searching the pages in the different stages. Currently, this functionality does not work, but in my understanding this is part of other issues (as well as adding a new paper). If this is the case, this comment can be ignored.

You are right, these are part of multiple other issues (e.g. #38) but should also be done soon in this sprint.

> I think the following problem can be handled in this issue: Because the value of a `PaperDetail` is displayed in a `textarea`, long titles (and maybe also other fields) can reach a length where a scrollbar appears. This looks very weird, and maybe just a line break should appear. I have no further suggestions since this can also become a separate issue since it is also effecting the paper details view of each separate paper.

Regarding this problem. I did not implement the `PaperDetail` but @Slartibartfass2. I wrote an issue for this bug and do not want to mix this bug fix with this new feature (see #339)

Commit:
ci: run svelte-kit sync in CI setup to resolve svelte aliases

ci: fix

Commit:
test: fix failing integration test after rebase

The tooltip for removing a paper from the reading list was different in the component and the test

mowi12 reviewed (changes requested):
I'm happy with the changes you made!
The only things you missed are the vertical scroll bar that appears, as soon as you open the paper details card.

And on my end, the integration tests do not pass. You changed the case of `Reading List` and probably forgot to change it in all tests.
For me, it was the `reading-list-entry.test.ts` file.

Lastly, it seems you have a bug.
If you click on a paper in the papers view, wait a second and click again, this click is counted as a double click instead of simply clicking again on the paper.

Commit:
fix: correctly handle timeout for differentation between single and double click

Commit:
fix: remove horizontal scrollbar when the paper details card is opened in the papers view

Merseleo commented:
> I'm happy with the changes you made! The only things you missed are the vertical scroll bar that appears, as soon as you open the paper details card.
> 
> And on my end, the integration tests do not pass. You changed the case of `Reading List` and probably forgot to change it in all tests. For me, it was the `reading-list-entry.test.ts` file.
> 
> Lastly, it seems you have a bug. If you click on a paper in the papers view, wait a second and click again, this click is counted as a double click instead of simply clicking again on the paper.

I hopefully fixed all problems. If you do not find anything and the CI runs without any fails, you can merge it on your own.

mowi12 reviewed (approved):
I found no problem with the current changes, nicely done!
